crates/autopilot/src/run_loop.rs

```diff
@@ -21,57 +25,105 @@ pub struct RunLoop {
+        solutions.shuffle(&mut rand::thread_rng());
+        solutions.sort_unstable_by(|left, right| left.1.objective.total_cmp(&right.1.objective));
 
-        Ok(())
+        // TODO: Keep going with other solutions until some deadline.
+        if let Some((index, solution)) = solutions.pop() {
+            tracing::info!("executing with solver {}", index);
+            match self
+                .execute(auction, id, &self.drivers[index], &solution)
+                .await
+            {
+                Ok(()) => tracing::info!("settled"),
+                Err(err) => {
+                    tracing::error!(?err, "solver {} failed to execute", err);
+                }
+            }
+        }
+
+        // TODO:
+        // - Think about what per auction information needs to be permanently stored. We might want
+        // to store the competition information and the full promised solution of the winner.
     }
 
-    async fn solve(&self, auction: &Auction, id: AuctionId) -> Vec<Result<solve::Response>> {
+    /// Returns the successful /solve responses and the index of the solver.
+    async fn solve(&self, auction: &Auction, id: AuctionId) -> Vec<(usize, solve::Response)> {
         let auction = solve::Auction {
             id,
             block: self.current_block.borrow().number,
             orders: auction.orders.iter().map(|_| solve::Order {}).collect(),
             prices: auction.prices.clone(),
         };
         let deadline = Utc::now() + chrono::Duration::from_std(SOLVE_TIME_LIMIT).unwrap();
-        let request = solve::Request { auction, deadline };
+        let request = &solve::Request { auction, deadline };
         let futures = self
             .drivers
             .iter()
-            .map(|driver| async {
-                match tokio::time::timeout(SOLVE_TIME_LIMIT, driver.solve(&request)).await {
-                    Ok(inner) => inner,
-                    Err(_) => Err(anyhow!("timeout")),
-                }
+            .enumerate()
+            .map(|(index, driver)| async move {
+                let result =
+                    match tokio::time::timeout(SOLVE_TIME_LIMIT, driver.solve(request)).await {
+                        Ok(inner) => inner,
+                        Err(_) => Err(anyhow!("timeout")),
+                    };
+                (index, result)
             })
             .collect::<Vec<_>>();
-        futures::future::join_all(futures).await
+        let results = futures::future::join_all(futures).await;
+        results
+            .into_iter()
+            .filter_map(|(index, result)| match result {
+                Ok(result) => Some((index, result)),
+                Err(err) => {
+                    tracing::warn!(?err, "driver {} solve error", err);
+                    None
+                }
+            })
+            .collect()
+    }
+
+    /// Execute the solver's solution. Returns Ok when the corresponding transaction has been mined.
```

Comment:
I think we should split `execute` and `wait` steps. Since, in the design we specify that solvers are allowed to reject execution (in case of price movement, meaning they know that their proposed solution wouldn't work). Specifically, I think the loop would look like:

```rust
let execution_promise = loop {
    let solution = match solution.pop() {
        Some(value) => value,
        None => break None,
     };
     
    match self
        .execute(auction, id, &self.drivers[index], &solution)
        .await
    {
        Ok(promise) => break Some(promise),
        Err(err) => {
            tracing::warn!(?err, "solver {} failed to execute", err);
            continue;
        }
    }
};
```

Reply:
Yes, will be something like this eventually.